### ISSUE TITLE
Pansharpening and TensorList

### ISSUE THREAD

**Issue by pierre-weiss** [deepinv/deepinv#163]
A few students have tried the PanSharpening forward model and it doesn't work. I wonder whether there is not a problem with the concatenation of operators with physics1 + physics2 as illustrated in 
https://deepinv.github.io/deepinv/auto_examples/basics/demo_pansharpening.html#sphx-glr-auto-examples-basics-demo-pansharpening-py 

When running the following lines : 

physics = dinv.physics.Pansharpen(
      img_size=img_size,
      device=device,
      noise_gray=dinv.physics.GaussianNoise(sigma=noise_level_img),
      noise_color=dinv.physics.GaussianNoise(sigma=noise_level_img)
)
data_fidelity = L2()
denoiser = dinv.models.DRUNet(pretrained="download").to(device)
prior = PnP(denoiser=denoiser)

model = optim_builder(
    iteration="PGD",
    prior=prior,
    data_fidelity=data_fidelity,
    early_stop=early_stop,
    max_iter=max_iter,
    verbose=verbose,
    params_algo=params_algo,
)

x_model, metrics = model(y, physics)

I get the following error in data_fidelity.py : 
    273         :return: (torch.tensor) gradient of the distance function :math:`\nabla_{u}\distance{u}{y}`.
    274         """
--> 275         return self.norm * (u - y)
    276 
    277     def prox_d(self, x, y, gamma=1.0):

The thing is you can't multiply a utils.nn.TensorList with a float (self.norm). So I am not sure that there is not a problem with the way operators are stacked. A solution could be to loop over the elements when using TensorLists, but this seems a bit tedious to change all the functions.

Any help is welcome !

**Comment by tachella** [comment]
Hi Pierre!

Many thanks for spotting this bug. Actually the problem was that the (left!) multiplication of a scalar times a `deepinv.utils.TensorList` was not implemented. We had only implemented the right multiplication.

I've just fixed this issue by adding the `__rmul__` method to the TensorList class (see [here](https://github.com/deepinv/deepinv/blob/ba666208bd1d84ce69e98a2a50dd8605c7e321ac/deepinv/utils/nn.py#L85)).

Now the following demo code works on my side (and gives very nice results out of the box!):

```
import deepinv as dinv

img_size = (3, 128, 128)
device = "cuda"
url = (
    "https://upload.wikimedia.org/wikipedia/commons/b/b4/"
    "Lionel-Messi-Argentina-2022-FIFA-World-Cup_%28cropped%29.jpg"
)
x = dinv.utils.load_url_image(url=url, img_size=(128, 128)).to(device)


noise_level_img = 0.1

physics = dinv.physics.Pansharpen(
img_size=img_size,
device=device,
noise_gray=dinv.physics.GaussianNoise(sigma=noise_level_img),
noise_color=dinv.physics.GaussianNoise(sigma=noise_level_img)
)

data_fidelity = dinv.optim.L2()
denoiser = dinv.models.DRUNet().to(device)
prior = dinv.optim.PnP(denoiser=denoiser)

model = dinv.optim.optim_builder(
    iteration="PGD",
    prior=prior,
    data_fidelity=data_fidelity,
    params_algo={"lambda": 1.0, "stepsize": 1.0, "g_param": 0.1},
)


y = physics(x)

x_model = model(y, physics)

dinv.utils.plot([y[0], y[1], x_model], ["RGB low-res", "grayscale hi-res", "reconstructed"])

```

**Comment by pierre-weiss** [comment]
Thanks a lot for the super quick fix!
Nice to better understand the basic mechanisms.

Pierre

Le lun. 26 févr. 2024, 22:44, Julián Tachella ***@***.***> a
écrit :
